* Feat Add AWS Cognito Authentication

JuanGuzman-io opened the pull request:
This pull request introduces several updates to enhance the authentication experience, improve code maintainability, and add new features. Key changes include the addition of a new `AuthCognitoComponent`, updates to the login page, and improvements to code readability and modularity.

### Authentication Enhancements:
* Added a new `AuthCognitoComponent` for handling Cognito-based authentication, including its HTML, SCSS, and TypeScript files. This component displays an authentication screen and manages login state using Cognito services. [[1]](diffhunk://#diff-d1ef656f7a63e680f451e8f14c034a0baa9d06ff4fffce53a379b1234130c4f2R1-R7) [[2]](diffhunk://#diff-b76e78344037ede225b54e921f83598d5605ebb36ca1d4ed27936f4ed55b227cR1-R47) [[3]](diffhunk://#diff-5ee3871457d1c2676fa2755cae01efe5cde6c0058d52a17737d9a4cc54b20908R1-R25) [[4]](diffhunk://#diff-5f880d5173292f1a4ae4501aa84c9509c804e8e6eed3b1f8370f88cae5b7931eR1-R24)
* Updated the `routes` in `app.routes.ts` to include a new path for the `AuthCognitoComponent`.

### Login Page Redesign:
* Revamped the `login.component.html` to include Azure AD login, password validation, and improved UI/UX with dynamic elements like loading spinners and password requirements.

### Code Quality Improvements:
* Made the `AlertComponent` constructor parameter `alertService` readonly and replaced null checks with optional chaining for better readability.
* Updated `alert.component.html` to handle nested error messages using optional chaining.

### Modularization:
* Converted `AssessorsChatWindowComponent` and `AppComponent` to standalone components by adding the `standalone: true` property and importing `CommonModule`. [[1]](diffhunk://#diff-9742c02232c2f313dce770b4394fc21ac3336c7825ee32aaa34dae2f0618557eL9-R12) [[2]](diffhunk://#diff-2ecac62bcc0d2e70e32d3c3009ac4cac6b6c3a6c1fe5e5f3929b1f5f809c1e9aR9-R21)

### Interface Updates:
* Added a new `LoginWithAzureAdRes` interface and an `error` property to the `MainResponse` interface in `responses.interface.ts` to support enhanced response handling. [[1]](diffhunk://#diff-40ff6abdd8899c8e51cec21cb2d1c255ceb012b563d7a0e5dcb5c3fbe50cf5c8R3) [[2]](diffhunk://#diff-40ff6abdd8899c8e51cec21cb2d1c255ceb012b563d7a0e5dcb5c3fbe50cf5c8R30-R37)

JuanGuzman-io committed:
✨ Integrate Cognito Microservice for authentication

Added AuthMicroserviceModule and AuthMicroserviceService to centralize authentication logic via an external microservice. Updated AuthService to delegate user authentication to the microservice, refactored affected modules to import the new module, and added dependencies for @nestjs/axios and axios. Included unit tests for the new service.

JuanGuzman-io committed:
♻️ Refactor user login response formatting

Replaces the previous user response with a formatted user object that includes a JWT token, formatted roles, config, and cycle. The password field is removed from the response for security. This improves the structure and security of the login response.

JuanGuzman-io committed:
✨ Add OAuth and password challenge support to AuthService

Introduces methods for OAuth authentication, including generating auth URLs, validating authorization codes, and completing password challenges. Refactors user authentication response logic into a reusable private method to support both standard and OAuth-based logins.

JuanGuzman-io committed:
🧪 Add and test new auth endpoints and improve login tests

Introduces new endpoints in AuthController for validating OAuth codes and completing password challenges, with corresponding unit tests. Expands AuthService tests to cover user not found and Cognito password challenge scenarios. Updates app controller test to check for QA API info in the root endpoint.

JuanGuzman-io committed:
🧪 Add comprehensive tests for CommentsController and CommentsService

Expanded unit tests for CommentsController to cover all controller methods using mocked service methods. Added detailed tests for CommentsService, including repository mocks and coverage for methods like getCommentsCount, getRawCommentsData, getRawCommentsExcel, getExcelComments, getQuickComments, getBatches, getCycles, and getActualCycle. These changes improve test coverage and reliability for the comments API.

JuanGuzman-io committed:
🧪 Add and expand unit tests for comments and evaluations

Added comprehensive error handling and success case tests for CommentsService and EvaluationsService, including edge cases and repository mock behaviors. Expanded EvaluationsController tests to cover all controller methods and their interactions with the service layer. These changes improve test coverage and reliability for comment and evaluation features.

JuanGuzman-io committed:
🧪 Add comprehensive tests for UsersController and UsersService

Expanded unit tests for UsersController to cover all main endpoints using a mocked UsersService. Added extensive tests for UsersService, including success and error scenarios for user CRUD operations, with all dependencies mocked. These changes improve test coverage and reliability for user-related logic.

JuanGuzman-io committed:
✨ Add endpoint to get OAuth provider authentication URL

Introduced a new GET /auth-url/:provider endpoint in AuthController to retrieve authentication URLs for OAuth providers. Also standardized string quotes to double quotes throughout the file for consistency.

JuanGuzman-io committed:
♻️ Refactor response format

Updated all import statements and string literals in auth.service.ts and auth-microservice.service.ts to use double quotes for consistency. Also made minor improvements to logging and response formatting for better clarity and maintainability.

JuanGuzman-io committed:
♻️ Refactor password challenge response formatting

Replaces manual response object with ResponseUtils.format for password change required flow. Also removes a console.log statement from getAuthURL.

JuanGuzman-io committed:
♻️ Remove test for Cognito new password challenge

Deleted the unit test that checks handling of the 'NEW_PASSWORD_REQUIRED' Cognito challenge in the AuthService. This may be due to changes in authentication flow or test coverage requirements.

xKeCo committed:
✨ feat(Login) P2-1589: added new login UI and connection with cognito auth services

xKeCo committed:
♻️ refactor(Login) P2-1589: update environment import in to promise service

xKeCo committed:
♻️ refactor(login) P2-1589: Update alert component and cognito service

xKeCo committed:
♻️ refactor(login): Update login component styles and set page title

xKeCo committed:
♻️ refactor(auth) P2-1589: Update POST_cognitoAuth method to use HttpClient and improve error handling in Cognito service

Cristian45 commented on client/public/img/qa-bg-1.jpg:
Por favor, consideremos almacenar esta imagen en S3. Este tema no es bloqueante; sin embargo, recomiendo revisarlo. Asimismo, tengamos en cuenta el peso de la imagen

Cristian45 commented on server/src/shared/microservice/auth-microservice/auth-microservice.service.spec.ts:
```diff
@@ -0,0 +1,387 @@
+  });
+
+  describe('createUser', () => {
+    it('✅ debe crear el usuario correctamente', async () => {
```
Por favor, poner este texto en ingles

Cristian45 commented on server/src/shared/microservice/auth-microservice/auth-microservice.service.spec.ts:
```diff
@@ -0,0 +1,387 @@
+      expect(result).toEqual(mockResponse.data);
+    });
+
+    it('❌ debe lanzar un error si la creación falla', async () => {
```
Por favor, pone este texto en ingles

JuanGuzman-io committed:
♻️ Translate test descriptions in createUser

Updated the test case descriptions in the createUser section from Spanish to English for consistency and clarity.

Cristian45 commented:
Approved ✅  
A plan must be created to:  
1. Store the image in S3 (considering its size).  
2. Apply custom Sonar configuration to avoid unit test code being counted as duplicated.
